### PR TITLE
Instrumentation: Report the size of plugin request

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

Adds reporting for plugin requests

**Why do we need this feature?**

For better instrumentation

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/observability-metrics-squad/issues/109

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 1]
```diff
@@ -29,12 +29,26 @@ var (
...
+		}, []string{"source", "plugin_id", "endpoint", "target"},
+	)
 )
 
 const (
 	statusOK        = "ok"
 	statusError     = "error"
 	statusCancelled = "cancelled"
+
+	EndpointCallResource   = "callResource"
```
Comment (wbrowne) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:48]: Nit: These can be unexported

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 2]
```diff
@@ -99,20 +113,25 @@ type Cfg struct {
...
+		string(cfg.Target)).Observe(requestSize)
+	return instrumentPluginRequest(ctx, cfg, req, EndpointCallResource, fn)
 }
 
 // InstrumentQueryDataRequest instruments success rate and latency of query data requests.
-func InstrumentQueryDataRequest(ctx context.Context, req *backend.PluginContext, cfg Cfg, fn func() error) error {
-	return instrumentPluginRequest(ctx, cfg, req, "queryData", fn)
+func InstrumentQueryDataRequest(ctx context.Context, req *backend.PluginContext, cfg Cfg,
+	requestSize float64, fn func() error) error {
+	pluginRequestSizeHistogram.WithLabelValues("backend-datasource", req.PluginID, EndpointQueryData,
```
Comment (wbrowne) [inline pkg/plugins/backendplugin/instrumentation/instrumentation.go:134]: Regarding the `source` label, I have some concerns about this as mentioned here https://github.com/grafana/grafana/pull/65527#discussion_r1158922049

File: pkg/plugins/manager/client/client.go [hunk 3]
```diff
@@ -82,14 +87,16 @@ func (s *Service) CallResource(ctx context.Context, req *backend.CallResourceReq
 		return fmt.Errorf("sender cannot be nil")
 	}
 
+	totalBytes := float64(len(req.Body))
```
Comment (wbrowne) [inline pkg/plugins/manager/client/client.go:90]: Nit: Move this to after line 95

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go (hunk 2 above)
Comment (LudoVio) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:134]: you can name it `grafana-backend`

File: pkg/plugins/backendplugin/instrumentation/instrumentation.go [hunk 4]
```diff
@@ -29,12 +29,26 @@ var (
...
 		Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 25, 50, 100},
 	}, []string{"plugin_id", "endpoint", "target"})
+
+	pluginRequestSizeHistogram = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "grafana",
+			Name:      "plugin_request_size_bytes",
+			Help:      "histogram of plugin request sizes returned",
+			Buckets:   []float64{128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576},
+		}, []string{"source", "plugin_id", "endpoint", "target"},
```
[the remark this answers is not in the document]
Comment (itsmylife) [reply pkg/plugins/backendplugin/instrumentation/instrumentation.go:39]: The status is a bit useless here. I am reporting the size of the request. What status should I report? It is going to be `ok` all the time.

**Review by wbrowne — APPROVED** [review]

**Pull request merged**